### ISSUE TITLE
Direct Consumer Offset Management

### ISSUE THREAD

**Issue by kevinconaway** [twmb/franz-go#89]
Most of our applications do not use consumer groups and are instead initialized with a static list of partitions to consume from.  I'm having trouble adapting our current usage to this library

How can "direct" consumers (those using `kgo.ConsumePartitions`) commit offsets?  Currently in `CommitOffsets`, the function exits early if it detects that the consumer is not in "group" mode. 

Separately, it doesn't appear that there is an option for `kgo.ConsumePartitions` to resume from the last committed offset, is that accurate?  I see that there is `SetOffsets` which _could_ be used if there was a way to fetch the last committed offsets for a group ID / partition but I don't see one.

**Comment by twmb** [comment]
The client doesn't itself support committing outside of the context of a group, but I think there are a few options to address this ticket.

1st option, available currently: you could use `kmsg.OffsetCommitRequest` directly to commit offsets outside of the context of a group. I think this still needs a group name (looking at the Kafka source), but I'm not 100% sure -- I've never not used group consumers when needing to commit offsets. The most tedious part of this is building the request, but `client.Request` handles issuing it properly and retrying, etc. This would work with `kmsg.ConsumePartitions` and directly consuming partitions.

2nd option, small bit of work: I could write a new group balancer, `ManualPartitionAssignor` or something, such that you can specify partitions manually, use group consuming, and the balancer will just use your exact assignments. This would allow you to use all of the automatic internal offset tracking / committing. I think this would make things easier in an application, but the main downside is this is actually using a group / requires a group ID.

Another option is to try to build some API for this into the client itself, but I think the former two options are far more compelling. What do you think? Also, I'm wondering what the use case is for committing offsets but not in a group / assigning partitions yourself externally (e.g. what reasoning could I give in "why to use this" if I add a `ManualPartitionAssignor`)?

---

> Separately, it doesn't appear that there is an option for kgo.ConsumePartitions to resume from the last committed offset, is that accurate?

That is accurate, _but_ you can use `kmsg.OffsetFetchRequest` to fetch the previous commits first, and then initialize a new client with the prior commits. It's a bit manual. Making this simpler would fall under the purview of #52, but I don't have the best ideas for that yet. Hopefully the ideas in the prior section would make this moot, though.

**Comment by kevinconaway** [comment]
Thanks so much for getting back to me.

>I'm wondering what the use case is for committing offsets but not in a group / assigning partitions yourself externally (e.g. what reasoning could I give in "why to use this" if I add a ManualPartitionAssignor)?

We have historically avoided using dynamic consumer groups as we have seen rebalance storms during kubernetes deployments & other operations.  This is detailed more in the initial confluent blog post for [static membership](https://www.confluent.io/blog/kafka-rebalance-protocol-static-membership/).  Unfortunately, we're not quite ready to use static membership on our end yet for a variety of reasons 

Thus we given static partition assignments to each consumer who _assign_ them on the consumer.  We do use the consumer group ID for offset management but we don't use dynamic group membership. This behavior is mentioned in passing in the [`KafkaConsumer`](https://kafka.apache.org/22/javadoc/org/apache/kafka/clients/consumer/KafkaConsumer.html) javadocs:

> It is also possible for the consumer to manually assign specific partitions (similar to the older "simple" consumer) using assign(Collection). In this case, dynamic partition assignment and consumer group coordination will be disabled.

**Comment by twmb** [comment]
I think the manual `kmsg.OffsetCommit` and `kmsg.OffsetFetch` bits would be the most applicable to your use case, although it is a little bit extra work. What do you think of this?

Alternatively, I think that the proposed manual partition assignor would avoid rebalance storms because it could be a cooperative balancer, meaning consuming would still occur during rebalances (so the rebalance wouldn't really be visible). IMO this would be pretty nifty to try.

**Comment by kevinconaway** [comment]
>I think the manual kmsg.OffsetCommit and kmsg.OffsetFetch bits would be the most applicable to your use case, although it is a little bit extra work. What do you think of this?

Its not ideal but it could work for the interim (or long term if you decide not to support this)

Perhaps it is my history with `confluent-kafka-go` that colors my perspective but ideally I would be able to pass in an offset like `kgo.NewOffset().Stored()` to `ConsumePartitions` and it would resume from the correct place.

For committing offsets, what would the impact be if `CommitOffsets` was modified to not require a group?  It looks like its on the `groupConsumer` now to use. the memberID/generationID but these can be blank for the "simple" consumer scenario per [the spec](https://cwiki.apache.org/confluence/display/KAFKA/A+Guide+To+The+Kafka+Protocol#AGuideToTheKafkaProtocol-OffsetCommit/FetchAPI)

>Note that when this API is used for a "simple consumer," which is not part of a consumer group, then the generationId must be set to -1 and the memberId must be empty (not null). Additionally, if there is an active consumer group with the same groupId, then the commit will be rejected (typically with an UNKNOWN_MEMBER_ID or ILLEGAL_GENERATION error).

**Comment by twmb** [comment]
`Stored` is a good idea, although to follow current naming I think I'd go with `AtCommitted`. This could be used with a new option, `DirectGroupManagement` or something (not 100% positive on name yet) to disable a consumer group, but internally save the consumer group name so that `SetOffsets` / `CommitOffsets` would work.

I'm not sure on the timeline for this -- conceptually the change is easy, but wiring in an offset fetch for direct partition consuming is at least a little bit of work. Could be a week or two. Until that time I think the current solution is to use OffsetFetch/OffsetCommit manually.

I don't think I'll add the manual partition balancer, then. Doesn't sound worth it to address this.

Out of curiosity, what has you checking out this library over your prior usage of confluent-kafka-go?

**Comment by kevinconaway** [comment]
> what has you checking out this library over your prior usage of confluent-kafka-go?

We've been trying to track down some heavy CPU usage in our applications that is attributed to `runtime.futex`.  Its my strong suspicion that this is related to repeatedly crossing the cgo barrier on producing/consuming.

> I'm not sure on the timeline for this 

No rush, I'm using some free time to pilot this out as a POC to see what impact it has for our use case.  I have a prototype working for `OffsetCommitRequest` and it is not too bad.

**Comment by twmb** [comment]
Thinking about this more, how easy is it to directly use `OffsetCommitRequest` / `OffsetFetchRequest`? I'm looking at this code for where it would be necessary to wire in an `OffsetFetchRequest` if an offset were `AtCommitted`, and it'd be an entirely new flow path, and it looks like it'd complicate the consumer.go `assignPartitions` code a good bit (which is already pretty complicated).

If it's easy enough to use `OffsetCommitRequest` / `OffsetFetchRequest`, I think my preferred path would be to add an example with those directly and call it good.

**Comment by kevinconaway** [comment]
Yes, some documentation would probably be sufficient here.  Once I read the protocol spec it was easy enough to figure out but I'm not sure that I should have needed to in order to use the library.  Is that fair?

**Comment by twmb** [comment]
Probably fair. I can have an example around it, and maybe mention it in the docs somewhere.

The kmsg documentation itself on the types is pretty thorough, but it's hard to discover what type you want if you don't know going into it.

**Comment by kevinconaway** [comment]
>The kmsg documentation itself on the types is pretty thorough

Yes, definitely.  Sorry, what I meant was confusing was _how_ to use those types.  For example, in my use case I don't need `MemberID`, `GenerationID` or `InstanceID` on the `OffsetCommitRequest` msg but I didn't really know what values to put there without reading the spec.

**Comment by twmb** [comment]
I see, yeah that would be great in an example. Also, if you aren't, be sure to use `kmsg.NewPtrOffset{Commit,Fetch}Request`, and use `New` functions for any struct -- these default appropriately just in case new fields are added in the future.

**Comment by twmb** [comment]
That's just as fine! NewPtr / New are just to provide whether you want to use `&req` vs. `req` at times.

**Comment by twmb** [comment]
I've added offset fetching and committing to a new kadm package:

https://github.com/twmb/franz-go/commit/e0d6ab19218824d3cf76a9a2041b68af9f247ae5#diff-e78537876fa4ebc174b5ceafc4b56453259679497f05261ccaf26d1e9eb3f22eR345-R450

The package is currently a separate module to avoid stabilization concerns. I'm going to play with this package a little bit, the aim is to merge it to franz-go proper asap (basically just dropping the go.mod and go.sum files). If you have a chance, I'd be curious if you could try it out and let me know what you think.

I'd also like to add an example of the kadm offset commit / fetch at some point.
